packages/react-select/src/Select.js

```diff
@@ -341,40 +342,59 @@ function toCategorizedOption(
   };
 }
 
-function buildCategorizedOptions(
-  props: Props,
-  selectValue: OptionsType
-): CategorizedGroupOrOption[] {
-  return (props.options
-    .map((groupOrOption, groupOrOptionIndex) => {
-      if (groupOrOption.options) {
-        const categorizedOptions = groupOrOption.options
-          .map((option, optionIndex) =>
-            toCategorizedOption(props, option, selectValue, optionIndex)
-          )
-          .filter(categorizedOption => isFocusable(props, categorizedOption));
-        return categorizedOptions.length > 0
-          ? {
-              type: 'group',
-              data: groupOrOption,
-              options: categorizedOptions,
-              index: groupOrOptionIndex,
-            }
+let buildCategorizedOptions = memoize(
+  function(props: Props, selectValue: OptionsType): CategorizedGroupOrOption[] {
+    return (props.options
+      .map((groupOrOption, groupOrOptionIndex) => {
+        if (groupOrOption.options) {
+          const categorizedOptions = groupOrOption.options
+            .map((option, optionIndex) =>
+              toCategorizedOption(props, option, selectValue, optionIndex)
+            )
+            .filter(categorizedOption => isFocusable(props, categorizedOption));
+          return categorizedOptions.length > 0
+            ? {
+                type: 'group',
+                data: groupOrOption,
+                options: categorizedOptions,
+                index: groupOrOptionIndex,
+              }
+            : undefined;
+        }
+        const categorizedOption = toCategorizedOption(
+          props,
+          groupOrOption,
+          selectValue,
+          groupOrOptionIndex
+        );
+        return isFocusable(props, categorizedOption)
+          ? categorizedOption
           : undefined;
-      }
-      const categorizedOption = toCategorizedOption(
-        props,
-        groupOrOption,
-        selectValue,
-        groupOrOptionIndex
-      );
-      return isFocusable(props, categorizedOption)
-        ? categorizedOption
-        : undefined;
-    })
-    // Flow limitation (see https://github.com/facebook/flow/issues/1414)
-    .filter(categorizedOption => !!categorizedOption): any[]);
-}
+      })
+      // Flow limitation (see https://github.com/facebook/flow/issues/1414)
+      .filter(categorizedOption => !!categorizedOption): any[]);
+  },
+  function(newArgs: any, lastArgs: any) {
+    const [newProps, newSelectValue] = (newArgs: [Props, OptionsType]);
+    const [lastProps, lastSelectValue] = (lastArgs: [Props, OptionsType]);
+
+    // Shallow compare props
+    // ignore menuIsOpen, as it's not used by sub-functions
+    for (let key in newProps) {
+      if (key === 'menuIsOpen') continue;
```

Comment:
Optimally I would only compare the props that are needed by this function, but the whole prop object is passed to sub-functions, so it would be to hard to keep track of.

I don't check `menuIsOpen` because it's not used here and is changing when I need memoization of this function.